ouroboros/chainsync/types.go

```diff
@@ -373,12 +552,235 @@ func (r Block) PointStruct() PointStruct {
+			ID:     r5.IntersectionFound.Tip.Hash,
+			Slot:   r5.IntersectionFound.Tip.Slot,
+		}
+		c.Intersection = r5.IntersectionFound.Point
+		c.Tip = &tip
+		c.Error = nil
+		c.ID = nil
+		return nil
+	} else if r5.IntersectionNotFound != nil {
+		// Emulate the v6 IntersectionNotFound error as best as possible.
+		tip := Tip{
+			Height: r5.IntersectionNotFound.Tip.BlockNo,
+			ID:     r5.IntersectionNotFound.Tip.Hash,
+		}
+		err := ResultError{Code: 1000, Message: "Intersection not found", Data: &tip}
+		c.Error = &err
+		return nil
+	}
+
+	// TODO: Further error handling here.
+	return nil
+}
+
+func (c CompatibleResultFindIntersection) String() string {
+	return fmt.Sprintf("intersection=[%v] tip=[%v] error=[%v] id=[%v]", c.Intersection, c.Tip, c.Error, c.ID)
+}
+
 // Covers all blocks except Byron-era blocks.
-type ResultPraos struct {
+type ResultNextBlockPraos struct {
 	Direction string `json:"direction,omitempty" dynamodbav:"direction,omitempty"`
 	Tip       *Tip   `json:"tip,omitempty"       dynamodbav:"tip,omitempty"`
 	Block     *Block `json:"block,omitempty"     dynamodbav:"block,omitempty"` // Forward
-	Point     Point  `json:"point,omitempty"     dynamodbav:"point,omitempty"` // Backward
+	Point     *Point `json:"point,omitempty"     dynamodbav:"point,omitempty"` // Backward
+}
+
+// Support findIntersect (v6) / FindIntersection (v5) universally.
+type CompatibleResultNextBlock ResultNextBlockPraos
+
+func (c *CompatibleResultNextBlock) UnmarshalJSON(data []byte) error {
+	// Assume v6 responses first, then fall back to manual v5 processing.
+	var r ResultNextBlockPraos
+	err := json.Unmarshal(data, &r)
+	if err == nil && r.Tip != nil {
+		*c = CompatibleResultNextBlock(r)
+		return nil
+	}
+
+	var r5 ResultNextBlockV5
+	err = json.Unmarshal(data, &r5)
+	if err != nil {
+		return err
+	} else if r5.RollForward != nil {
+		tip := Tip{Height: r5.RollForward.Tip.BlockNo, ID: r5.RollForward.Tip.Hash, Slot: r5.RollForward.Tip.Slot}
+
+		txArray := make([]Tx, len(r5.RollForward.Block.Body))
+		for _, t := range r5.RollForward.Block.Body {
+			withdrawals := make(map[string]Lovelace)
+			for txid, amt := range t.Body.Withdrawals {
+				withdrawals[txid] = Lovelace{Lovelace: num.Int64(amt), Extras: nil}
+			}
+
+			tx := Tx{
+				ID:                       t.ID,
+				Spends:                   t.InputSource,
+				Inputs:                   t.Body.Inputs,
+				References:               t.Body.References,
+				Collaterals:              t.Body.Collaterals,
+				TotalCollateral:          t.Body.TotalCollateral,
+				CollateralReturn:         (*TxOut)(t.Body.CollateralReturn),
+				Outputs:                  t.Body.Outputs,
+				Certificates:             t.Body.Certificates,
+				Withdrawals:              withdrawals,
+				Fee:                      Lovelace{Lovelace: t.Body.Fee, Extras: nil},
+				ValidityInterval:         t.Body.ValidityInterval,
+				Mint:                     nil, // TODO - Differences appear to be too much to handle.
```

Comment:
This is one that we'll definitely need to handle, we depend on the mint field in plenty of cases

Reply:
So, `mint` is weird. I'm going to update the unit tests but here are examples of what we see in v5 and v6.

v5:
```
"mint": {
    "coins": 0,
    "assets": {
        "3542acb3a64d80c29302260d62c3b87a742ad14abf855ebc6733081e.00": 3,
        "e0a714319812c3f773ba04ec5d6b3ffcd5aad85006805b047b082541": 4,
        "e0a714319812c3f773ba04ec5d6b3ffcd5aad85006805b047b082541.00": -3
    }
}
```

v6:
```
"mint": {
    "be8dc1ca8b6b1735175a24b3b3c39d7327aba843c16a9b99924f3476": {
        "a3589dbc6194521a": 98371879850696154
    }
}
```

Looking at the API specs, it looks like v5 is `POLICY_ID.ASSET_NAME`, and v6 is just "property names." I'm stumped on this one.

Reply:
It looks like `chainsync.Value` has been updated from having a map from AssetId to amount (AssetId being `PolicyID.AssetNameHex`), to now being:
```
map[string]map[string]num.Int
```

i.e. a map from policy ID, to a map from asset name, to amount.

Reply:
Thanks. As I test `sundae-sync`, I'll fold in any relevant changes in the next PR.